lighthouse-core/test/audits/dobetterweb/no-vulnerable-libraries-test.js

```diff
@@ -99,3 +99,20 @@ describe('Avoids front-end JavaScript libraries with known vulnerabilities', ()
     assert.equal(auditResult.score, 1);
   });
 });
+
+describe('Every snyk vulnerability has an upperbound', () => {
+  for (const vulns of Object.values(NoVulnerableLibrariesAudit.snykDB.npm)) {
+    for (const vuln of vulns) {
+      for (const semver of vuln.semver.vulnerable) {
+        assert.notEqual(semver, '*', 'invalid semver: * is not allowed');
+
+        const clauses = semver.split('||');
+        for (const clause of clauses) {
+          if (!clause.trim().startsWith('=') && !clause.includes('<')) {
```

Comment:
maybe give a few examples here for easier parsing? I had to go checkout the snapshot to see why it's `includes('<')` but `.trim().startsWith('=')`

Reply:
> I had to go checkout the snapshot to see why it's `includes('<')` but `.trim().startsWith('=')`

hmm, I still don't understand why that's the case :).

Isn't something like `'=5.0 >=10'` valid with no upper bound? But it wouldn't fail this check. Or `'<5.0 >=10'`? Maybe I just don't understand what we're checking here :)

Reply:
well those aren't real ranges :)

I suppose it's a meta question of if we should be checking to see that they're valid ranges at all

Reply:
> well those aren't real ranges :)

why aren't they? Isn't `'=5.0 >=10'` just saying 5.x or anything over 10 is vulnerable?

Reply:
> well those aren't real ranges :)

this

> valid ranges at all

doesn't make sense to validate snyk like that imo

Reply:
`semver.validRange('=5.0 >=10')` is good with it, so maybe explain it to me :)

Reply:
a space is an `AND`

`OR` is either `||` a separate entry in the `vulnerable` array

so while it might technically be valid, nothing will ever satisfy the range `'=5 >=10'`

Reply:
semver package's `validRange` just verifies the string conforms to the grammar defined here: https://github.com/npm/node-semver/blob/ce6190e2b681700dcc5d7309fe8eda99941f712d/range.bnf not that the range is satisfiable.

Reply:
> so while it might technically be valid, nothing will ever satisfy the range '=5 >=10'

Ah, I see your point, but it could also be `'=11 >= 10'`. I assume most of these are auto generated, so it doesn't seem out of the realm of possibility. See my other comment for one possibility rather than getting clever and hoping these ranges always make sense :)

Reply:
> Ah, I see your point, but it could also be '=11 >= 10'

oh, no, wait, that's wrong. So I'll have to fall back to hand waving about invalid ranges :)

Reply:
> doesn't make sense to validate snyk like that imo

didnt realize how easy it'd be to use `semver` to validate the ranges, sooo nvm. it's done indirectly with this test now.